src/main/java/com/exasol/spark/s3/ExasolBatchWrite.java

```diff
@@ -0,0 +1,163 @@
+package com.exasol.spark.s3;
+
+import static com.exasol.spark.s3.Constants.INTERMEDIATE_DATA_PATH;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.spark.sql.connector.write.*;
+
+import com.exasol.errorreporting.ExaError;
+
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+/**
+ * An Exasol {@link BatchWrite} implementation.
+ */
+public class ExasolBatchWrite implements BatchWrite {
+    private static final Logger LOGGER = Logger.getLogger(ExasolBatchWrite.class.getName());
+
+    private final ExasolOptions options;
+    private final BatchWrite delegate;
+
+    /**
+     * Creates a new instance of {@link ExasolBatchWrite}.
+     *
+     * @param options  user provided options
+     * @param delegate delegate {@code CSV} batch write
+     */
+    public ExasolBatchWrite(final ExasolOptions options, final Write delegate) {
+        this.options = options;
+        this.delegate = delegate.toBatch();
+    }
+
+    @Override
+    public DataWriterFactory createBatchWriterFactory(final PhysicalWriteInfo info) {
+        return delegate.createBatchWriterFactory(info);
+    }
+
+    @Override
+    public boolean useCommitCoordinator() {
+        return delegate.useCommitCoordinator();
+    }
+
+    @Override
+    public void abort(final WriterCommitMessage[] messages) {
+        LOGGER.info("Running abort stage of the job.");
+        cleanup();
+        delegate.abort(messages);
+    }
+
+    private void cleanup() {
+        LOGGER.info(() -> "Running cleanup process for directory '" + this.options.get(INTERMEDIATE_DATA_PATH) + "'.");
+    }
+
+    @Override
+    public void commit(final WriterCommitMessage[] messages) {
+        LOGGER.info("Committing the file writing stage of the job.");
+        delegate.commit(messages);
+        prepareIntermediateData();
```

Comment:
Yes, that is fine. We commit the delegate (which CSV writer). But this method is for the connector job.

From the [contact of the method](https://github.com/apache/spark/blob/v3.4.0/sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java#L79-L81):
> If this method fails (by throwing an exception), this writing job is considered to to have been failed, and [abort(WriterCommitMessage[])](https://spark.apache.org/docs/3.2.1/api/java/org/apache/spark/sql/connector/write/BatchWrite.html#abort-org.apache.spark.sql.connector.write.WriterCommitMessage:A-) would be called.

If the `commit` method itself fails via throwing exception, that means we were not able to import successfully. And it will go into the abort method where the cleanup will be run.

Reply:
But good point, maybe I should also set the autocommit to false on Exasol connection and commit once import is successfully run.

Reply:
So, can I think of this as a staged commit? Each stage of the pipelines first tries to commit the previous (here CSVWriter) and then tries to commit for itself? Would actually make sense for dataflow, because different parts could succeed while others may fail.

Reply:
Yes, that is what I understood. The delegate will wait for all of them (all partitions are committed) and then commit itself with this method.

We also keep the `useCommitCoordinator` to true to wait for all of the and not speculatively commit.

Reply:
Created #162, for later to add the discussion to the developer guide.